File: backend/app/harness/pipeline/phases/load_confirmed_keywords.py
"""LoadConfirmedKeywordsPhase — read the user-confirmed keyword plan from
Redis (set by /rounds/{id}/confirm-keywords) and merge it into QueryPlan.

Mutates ``ctx.get('plan_query')['query_plan']`` in place: this is intentional
— LoadConfirmedKeywords is a strict overlay on PlanQuery's output, not a
parallel branch. Returns the per-source query map and dynamic synonyms for
downstream phases."""
from __future__ import annotations

import logging

from app.harness.pipeline.types import RoundContext

logger = logging.getLogger(__name__)


class LoadConfirmedKeywordsPhase:
    name = "load_confirmed_keywords"
    deps = ["plan_query"]
    progress_range = (0.20, 0.21)
    can_interrupt = False

    async def execute(self, ctx: RoundContext) -> dict:
        import json

        import redis.asyncio as aioredis

        from app.config import settings

        per_source_queries: dict | None = None
        dynamic_synonyms: dict | None = None

        if not settings.enable_per_source_keywords:
            return {"per_source_queries": None, "dynamic_synonyms": None}

        query_plan = ctx.get("plan_query")["query_plan"]
        try:
            r = aioredis.from_url(settings.redis_url)
            try:
                raw = await r.get(f"keyword_plan:{ctx.round_id}")
                if not raw:
                    return {"per_source_queries": None, "dynamic_synonyms": None}
                plan_data = json.loads(raw)
                if plan_data.get("confirmed"):
                    per_source_queries = {}
                    for p in plan_data.get("source_plans", []):
                        if not p.get("enabled", True):
                            continue
                        sid = p["source_id"]
                        per_source_queries[sid] = {
                            "complex": p.get("query") or "",
                            "medium": p.get("query_medium") or "",
                            "simple": p.get("query_simple") or "",
                        }
                    if per_source_queries:
                        # Respect user's enabled set verbatim — see comment in
                        # legacy code re: not intersecting with agent plan
                        # (would drop sources the user explicitly enabled).
                        query_plan.sources = list(per_source_queries.keys())
                        logger.info(
                            "[PerSourceKW] %d sources confirmed (3-tier)",
                            len(per_source_queries),
                        )
                    if "base_query" in plan_data:
                        query_plan.base_query = plan_data["base_query"]
                        query_plan.expanded_terms = [
                            w for w in plan_data["base_query"].split() if len(w) >= 2
                        ]
                        logger.info(
                            "[UserOverride] base_query → %s",
                            query_plan.base_query[:60],
                        )
                    for k in (
                        "original_chinese_query", "exclude_terms",
                        "year_from", "year_to", "max_per_source",
                        "language_scope",
                    ):
                        if k in plan_data:
                            attr = "max_results_per_source" if k == "max_per_source" else k
                            setattr(query_plan, attr, plan_data[k])
                dynamic_synonyms = plan_data.get("synonyms")
                if dynamic_synonyms:
                    logger.info(
                        "[PerSourceKW] %d synonym groups loaded",
                        len(dynamic_synonyms),
                    )
            finally:
                await r.close()
        except Exception as e:
            logger.warning("[PerSourceKW] failed to load confirmed plan: %s", e)

        return {
            "per_source_queries": per_source_queries,
            "dynamic_synonyms": dynamic_synonyms,
        }

File: client/src/data/pipeline/phases/loadConfirmedKeywords.ts
/**
 * LoadConfirmedKeywordsPhase — 读用户在 awaiting_keywords 阶段确认/编辑过的 keyword plan
 * 并 merge 进 ctx.queryPlan。
 *
 * 移植自 backend `phases/load_confirmed_keywords.py:17-95`，差异：
 * - 删除 redis 依赖；keyword plan 直接存 round.search_queries_json 的 `keyword_plan` 子字段
 *   （由 RoundOrchestrator.confirmKeywords() 写入）
 * - 走 SQLite roundRepo 读
 * - 用户没确认时（status 不是 awaiting_keywords→searching），返回 null（视同 PRD agentic 输出）
 */
import { getRound } from '@/data/sqlite/repos/roundRepo'

import type { RoundContext } from '../context'
import type { Phase } from '../runner'

import type { PlanQueryOutput } from './planQuery'

export interface LoadConfirmedKeywordsOutput {
  /** key=source, val={complex/medium/simple 三档 query 字符串} */
  perSourceQueries: Record<string, { complex: string; medium: string; simple: string }> | null
  dynamicSynonyms: Record<string, string[]> | null
}

export const loadConfirmedKeywordsPhase: Phase = {
  name: 'load_confirmed_keywords',
  deps: ['plan_query'],
  progressRange: [0.20, 0.21] as const,

  async execute(ctx: RoundContext): Promise<LoadConfirmedKeywordsOutput> {
    const round = await getRound(ctx.roundId)
    const plan = ctx.get<PlanQueryOutput>('plan_query')

    const sq = round?.search_queries
    if (!sq || typeof sq !== 'object') {
      return { perSourceQueries: null, dynamicSynonyms: null }
    }
    const planRecord = sq as Record<string, unknown>
    const kp = planRecord.keyword_plan as Record<string, unknown> | undefined
    if (!kp || !kp.confirmed) {
      return { perSourceQueries: null, dynamicSynonyms: null }
    }

    let perSourceQueries: LoadConfirmedKeywordsOutput['perSourceQueries'] = null
    const sourcePlans = Array.isArray(kp.source_plans) ? (kp.source_plans as Array<Record<string, unknown>>) : []
    if (sourcePlans.length > 0) {
      perSourceQueries = {}
      for (const p of sourcePlans) {
        if (p?.enabled === false) continue
        const sid = String(p.source_id ?? '')
        if (!sid) continue
        perSourceQueries[sid] = {
          complex: String(p.query ?? '').trim(),
          medium: String(p.query_medium ?? '').trim(),
          simple: String(p.query_simple ?? '').trim(),
        }
      }
      if (Object.keys(perSourceQueries).length > 0) {
        // 替换 plan.perSource keys 为用户的 enabled set（verbatim，不做 intersection）
        const replaced: typeof plan.queryPlan.perSource = {}
        for (const sid of Object.keys(perSourceQueries)) {
          replaced[sid] = plan.queryPlan.perSource[sid]
            ?? {
              keywords: perSourceQueries[sid].complex.split(/\s+/).filter((w) => w.length >= 2),
              filters: { ...plan.queryPlan.meta },
            }
        }
        plan.queryPlan.perSource = replaced
      } else {
        perSourceQueries = null
      }
    }

    if (typeof kp.base_query === 'string') {
      plan.queryPlan.meta.baseQuery = kp.base_query
    }
    for (const k of ['year_from', 'year_to', 'language_scope', 'exclude_terms'] as const) {
      if (k in kp) {
        ;(plan.queryPlan.meta as Record<string, unknown>)[
          k === 'language_scope'
            ? 'languageScope'
            : k === 'year_from'
              ? 'yearFrom'
              : k === 'year_to'
                ? 'yearTo'
                : 'excludeTerms'
        ] = kp[k]
      }
    }

    const dynamicSynonyms = kp.synonyms && typeof kp.synonyms === 'object'
      ? (kp.synonyms as Record<string, string[]>)
      : null

    return { perSourceQueries, dynamicSynonyms }
  },
}
